subprojects/core/src/main/java/org/gradle/api/internal/file/archive/AbstractArchiveFileTree.java

```diff
@@ -21,10 +21,36 @@
 import org.gradle.api.internal.tasks.TaskDependencyContainer;
 import org.gradle.api.internal.tasks.TaskDependencyResolveContext;
 import org.gradle.api.provider.Provider;
+import org.gradle.cache.CacheRepository;
+import org.gradle.cache.FileLockManager;
+import org.gradle.cache.PersistentCache;
+import org.gradle.cache.internal.CacheFactory;
+import org.gradle.cache.internal.CacheScopeMapping;
+import org.gradle.cache.internal.DefaultCacheRepository;
+import org.gradle.cache.internal.scopes.DefaultCacheScopeMapping;
+import org.gradle.initialization.GradleUserHomeDirProvider;
+import org.gradle.initialization.layout.GlobalCacheDir;
+import org.gradle.util.GradleVersion;
 
 import java.io.File;
 
+import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
+
 public abstract class AbstractArchiveFileTree implements FileSystemMirroringFileTree, TaskDependencyContainer {
+    private static final String EXPANSION_CACHE_KEY = "compressed-file-expansion";
+    private static final String EXPANSION_CACHE_NAME = "Compressed Files Expansion Cache";
+
+    protected final PersistentCache expansionCache;
+
+    protected AbstractArchiveFileTree(CacheFactory cacheFactory, GradleUserHomeDirProvider userHomeDirProvider) {
+        CacheScopeMapping cacheScopeMapping = new DefaultCacheScopeMapping(new GlobalCacheDir(userHomeDirProvider).getDir(), GradleVersion.current());
+        CacheRepository cacheRepository = new DefaultCacheRepository(cacheScopeMapping, cacheFactory);
```

Comment:
I think this is injectable as a service in the daemon. It isn't (or wasn't) available in the worker process, so instead of carrying around `CacheFactory` and `GradleUserHomeDirProvider`, I think you should be able to get an existing `CacheRepository` service.

Reply:
I'll give this a try.